sdk/src/impl/position-impl.ts

```diff
@@ -315,17 +362,57 @@ export class PositionImpl implements Position {
     );
 
     const accountExemption = await this.ctx.fetcher.getAccountRentExempt();
-    const { ataTokenAddresses, resolveAtaIxs } = await resolveAtaForMints(this.ctx, {
-      mints: initializedRewards.map((r) => r.mint),
-      accountExemption,
-      receiver: destinationWalletKey,
-      payer: ataPayerKey,
-    });
+    const affiliatedMints = getTokenMintsFromWhirlpools([whirlpool]);
+    const walletTokenAccountsByMint = { ...ownerTokenAccountsRecord };
+
+    const solInRewardMints = initializedRewards.some((info) => TokenUtil.isNativeMint(info.mint));
+    const solInWalletTokenAccounts = !!walletTokenAccountsByMint[NATIVE_MINT.toBase58()];
+    const mintsToResolveAtasFor = affiliatedMints.filter(
+      // Filter out mints for which we already have token accounts
+      (mint) => !walletTokenAccountsByMint[mint.toBase58()]
+    );
+
+    if (mintsToResolveAtasFor.length > 0) {
+      const { ataTokenAddresses, resolveAtaIxs } = await resolveAtaForMints(this.ctx, {
+        mints: mintsToResolveAtasFor,
+        accountExemption,
+        receiver: destinationWalletKey,
+        payer: ataPayerKey,
+      });
+
+      for (const mint of Object.keys(ataTokenAddresses)) {
+        walletTokenAccountsByMint[mint] = ataTokenAddresses[mint];
+      }
 
-    if (resolveATA) {
       txBuilder.addInstructions(resolveAtaIxs);
     }
 
+    let unwrapSolIx: Instruction | undefined;
+    if (solInRewardMints && !solInWalletTokenAccounts) {
+      const { wSolAccount, wrapIx, unwrapIx } = wrapSOL(
+        destinationWalletKey,
+        ZERO,
+        accountExemption,
+        ataPayerKey,
+        destinationWalletKey
+      );
+
+      walletTokenAccountsByMint[NATIVE_MINT.toBase58()] = wSolAccount;
+      txBuilder.addInstruction(wrapIx);
+      unwrapSolIx = unwrapIx;
+    }
+
+    const tokenOwnerAccountA = walletTokenAccountsByMint[whirlpool.tokenMintA.toBase58()];
+    invariant(
+      !!tokenOwnerAccountA,
+      `No owner token account provided for wallet ${destinationWalletKey.toBase58()} for token A ${whirlpool.tokenMintA.toBase58()}`
+    );
+    const tokenOwnerAccountB = walletTokenAccountsByMint[whirlpool.tokenMintB.toBase58()];
+    invariant(
+      !!tokenOwnerAccountB,
+      `No owner token account provided for wallet ${destinationWalletKey.toBase58()} for token B ${whirlpool.tokenMintB.toBase58()}`
+    );
```

Comment:
collectrewards needs to verify reward mint token accounts, not the pool token accounts.